cmd/bolt/main.go

```diff
@@ -416,9 +419,155 @@ func (cmd *DumpCommand) PrintPage(w io.Writer, r io.ReaderAt, pageID int, pageSi
 // Usage returns the help message.
 func (cmd *DumpCommand) Usage() string {
 	return strings.TrimLeft(`
-usage: bolt dump -page PAGEID PATH
+usage: bolt dump PATH pageid [pageid...]
 
-Dump prints a hexadecimal dump of a single page.
+Dump prints a hexadecimal dump of one or more pages.
+`, "\n")
+}
+
+// DumpItemCommand represents the "dump-item" command execution.
+type DumpItemCommand struct {
+	Stdin  io.Reader
+	Stdout io.Writer
+	Stderr io.Writer
+}
+
+// newDumpItemCommand returns a DumpItemCommand.
+func newDumpItemCommand(m *Main) *DumpItemCommand {
+	return &DumpItemCommand{
+		Stdin:  m.Stdin,
+		Stdout: m.Stdout,
+		Stderr: m.Stderr,
+	}
+}
+
+type dumpItemOptions struct {
+	help      bool
+	keyOnly   bool
+	valueOnly bool
+}
+
+// Run executes the command.
+func (cmd *DumpItemCommand) Run(args ...string) error {
+	// Parse flags.
+	options := &dumpItemOptions{}
+	fs := flag.NewFlagSet("", flag.ContinueOnError)
+	fs.BoolVar(&options.keyOnly, "key-only", false, "Only print the key.")
+	fs.BoolVar(&options.valueOnly, "value-only", false, "Only print the value.")
+	fs.BoolVar(&options.help, "h", false, "")
+	if err := fs.Parse(args); err != nil {
+		return err
+	} else if options.help {
+		fmt.Fprintln(cmd.Stderr, cmd.Usage())
+		return ErrUsage
+	}
+
+	if options.keyOnly && options.valueOnly {
+		return fmt.Errorf("The --key-only or --value-only flag may be set, but not both.")
+	}
+
+	// Require database path and page id.
+	path := fs.Arg(0)
+	if path == "" {
+		return ErrPathRequired
+	} else if _, err := os.Stat(path); os.IsNotExist(err) {
+		return ErrFileNotFound
+	}
+
+	// Read page id.
+	pageID, err := strconv.Atoi(fs.Arg(1))
+	if err != nil {
+		return err
+	}
+
+	// Read item id.
+	itemID, err := strconv.Atoi(fs.Arg(2))
+	if err != nil {
+		return err
+	}
+
+	// Open database file handler.
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = f.Close() }()
+
+	// Retrieve page info and page size.
+	_, buf, err := ReadPage(path, pageID)
+	if err != nil {
+		return err
+	}
+
+	if !options.valueOnly {
```

Comment:
```go
if !options.valuesOnly { ... }
if !options.keyOnly { ... }
```
so the output with no flags exactly matches the physical layout since this is a raw dump.

etcdctl's default output mode inserts a newline and in retrospect I think it makes the output really difficult to parse, but it's too late now. `printf("%q\n%q\n",key,val)` seems like the best/easiest human-readable way to print binary data, if that's worth pursuing in the future.

Reply:
How about we do something like this:

```
bolt page-item <dbfile> <pageid> <itemid> 
"<%q encoded key>"
"<%q encoded value>"
```

```
bolt page-item --raw-key-only <dbfile> <pageid> <itemid>
<key bytes>
```

```
bolt page-item --raw-value-only <dbfile> <pageid> <itemid>
<value bytes>
```
?

This solves a couple problems:

- It provides a default output that won't mess up someones terminal by outputting raw bytes, and that prints out the key and value, which, if ASCII, will be rendered nicely.
- It avoid use of "\n" as a delimiter between binary output.
- If one needs the raw bytes, they have convenient flags they can use to get the key and value individually.

?